src/CachePlugin.php

```diff
@@ -69,15 +69,42 @@ public function handleRequest(RequestInterface $request, callable $next, callabl
         $cacheItem = $this->pool->getItem($key);
 
         if ($cacheItem->isHit()) {
-            // return cached response
             $data = $cacheItem->get();
-            $response = $data['response'];
-            $response = $response->withBody($this->streamFactory->createStream($data['body']));
+            if (isset($data['expiresAt']) && time() < $data['expiresAt']) {
+                // This item is still valid according to previous cache headers
+                return new FulfilledPromise($this->createResponseFromCacheItem($cacheItem));
+            }
+
+            // Add headers to ask the server if this cache is still valid
+            if ($modifiedSinceValue = $this->getModifiedSinceHeaderValue($cacheItem)) {
+                $request = $request->withHeader('If-Modified-Since', $modifiedSinceValue);
+            }
 
-            return new FulfilledPromise($response);
+            if ($etag = $this->getETag($cacheItem)) {
+                $request = $request->withHeader('If-None-Match', $etag);
+            }
         }
 
         return $next($request)->then(function (ResponseInterface $response) use ($cacheItem) {
+            if (304 === $response->getStatusCode()) {
+                if (!$cacheItem->isHit()) {
```

Comment:
would this be because the userland code did an If-None-Match / If-Modified-Since header? but indeed i think this is the only thing we can do at this point.